in-progress/7228-unified-bbjs-interface.md

```diff
@@ -0,0 +1,243 @@
+- Simplify the installation process for the user (downloading the native extension would be handled by the bb.js npm package)
+
+## Usage scenarios
+
+We expect to use this API for proving and verifying:
+
+- In a node.js app (like Aztec's prover)
+- In a webapp
+- Via the bb.js CLI for local Noir development
+
+## Approach
+
+The bb wasm API and the bb CLI API are different in nature. The wasm API is more fine-grained, since it can make use of a long-lived bb instance that caches compiled circuits and keys. It also exposes functions unrelated to proving, such as hashing, signing, and other cryptographic primitives. On the other hand, the CLI API bundles together common proving operations, as it is stateless and it benefits from reducing the total number of calls for each operation.
+
+Since we don't have plans on reworking bb as a long-running native process now, we are limited to communicate with native bb via its CLI, so we'll shoot for a coarse API. This should also be simpler for users to work with. Eventually, if we implement the long-running native bb process, we can provide a more fine-grained API, and reimplement the one in this doc via multiple calls to it.
+
+For the design, we gather "inspiration" from the native CLI, the bb.js CLI, the noir-js-bb-backend, and the Aztec bb-prover.
+
+## Interface and implementation
+
+We propose the following interface, to be implemented via the native CLI and wasm as described below.
+
+### Types
+
+#### `Circuit`
+
+Circuit bytecode in ACIR format.
+
+#### `Witness`
+
+Binary witness as outputted by the ACVM simulator.
+
+#### ~~`PathOrData`~~
+
+Either a path to the filesystem or the actual contents. This would allow us to propagate paths to disk from the bb.js interface into the native bb, instead of having to read a file into memory from bb.js just to conform to the shared interface, and then write again to disk to call the native bb CLI.
+
+:question: We discarded this type for simplicity reasons, but there are reasons to keep it. Even after we discard the native CLI, we may still want to just point native bb to disk as a means to share data between bb.js and bb, especially if we cache more aggressively.
+
+#### `Flavor`
+
+Specifies the proving system for the proof. Not all methods may be supported for all flavors, either because of limitations (eg AVM is not supported in wasm) or because it doesn't make sense (eg there is a single AVM verification key). Valid values:
+
+- `UltraPlonk`
+- `UltraHonk`
+- `MegaHonk`
+- `ClientIVC`
+- `AVM`
+
+:question: I collected these values from scanning the bb CLI interface. Are they correct? Should we only target a subset of them? Is "flavor" the correct term? Should we spell it properly as "flavour"?
+
+### Interface
+
+#### `prove(circuit: Circuit, witness: Witness, flavor: Flavor): { proof: Buffer }`
```

Comment:
Note that proving the AVM takes quite different inputs (and works different from) other circuits. Ref `bb_prover.ts` and `bb/execute.ts`; this interface doesn't fit as far as I can see. 

Likely same for verification (we now currently pass things from the prover by prepending to the VK, but we'll eventually have to pass public inputs directly I think).